torch/csrc/profiler/collection.cpp

```diff
@@ -822,147 +822,225 @@ trace_ptr_t addKinetoEvents(
+              insert_tensor(p.metadata_);
+              if (p.grad_metadata_.has_value()) {
+                insert_tensor(*p.grad_metadata_);
+              }
+              for (auto& state_i : p.state_) {
+                insert_tensor(state_i.second);
               }
             }
-          },
-          [](const auto&) {}));
+          }
+        },
+        [](const auto&) {}));
+  }
+
+  // Step 2) Map TensorImpls to allocations and remove non-Tensor allocations.
+  // --------------------------------------------------------------------------
+  {
+    ska::flat_hash_map<StorageImplData, TensorImplAddress> current;
+    auto update = [&current](RawTensorStoragePair& t, PairType update_type) {
+      auto it = current.insert({t.storage_.first, NoTensorImpl}).first;
+      if (t.impl_.first != NoTensorImpl) {
+        it->second = t.impl_.first;
+      }
+
+      if (t.type_ == update_type) {
+        t.impl_.first = it->second;
+      }
+
+      if (t.type_ == PairType::kAllocation || t.type_ == PairType::kFree) {
+        current.erase(it);
+      }
+    };
+
+    for (auto& t : raw_tensors) {
+      update(t, PairType::kFree);
+    }
+
+    current.clear();
+    for (auto it = raw_tensors.rbegin(); it != raw_tensors.rend(); ++it) {
+      update(*it, PairType::kAllocation);
     }
 
-    // Handle any allocation events which we cannot prove are for
-    // `StorageImpl`s.
-    tensors.erase(
+    raw_tensors.erase(
         std::remove_if(
-            tensors.begin(),
-            tensors.end(),
-            [&tensor_set](const auto& i) {
-              return tensor_set.find(i.storage_id_) == tensor_set.end();
+            raw_tensors.begin(),
+            raw_tensors.end(),
+            [](const auto& t) {
+              return t.impl_.first == NoTensorImpl ||
+                  t.storage_.first == StorageImplData{nullptr};
             }),
-        tensors.end());
+        raw_tensors.end());
+  }
+
+  // Step 3) Handle ABA for Storage.
+  // --------------------------------------------------------------------------
+  ska::flat_hash_map<versioned_storage_t, size_t, PairHash> storage_ids;
+  {
+    ska::flat_hash_map<StorageImplData, size_t> storage_version;
+    for (auto& t : raw_tensors) {
+      auto& id = storage_version.insert({t.storage_.first, 0}).first->second;
+      t.storage_.second = id;
+      storage_ids.insert({t.storage_, storage_ids.size()});
+      id += (t.type_ == PairType::kFree);
+    }
+  }
+
+  // Step 4) Gather all (unversioned) TensorImpls for a given versioned storage.
+  // --------------------------------------------------------------------------
+  ska::flat_hash_map<
+      versioned_storage_t,
+      ska::flat_hash_set<TensorImplAddress>,
+      PairHash>
+      lookahead;
+  {
+    for (auto& t : raw_tensors) {
+      lookahead[t.storage_].insert(t.impl_.first);
+    }
   }
 
-  // Step 2) Handle the case that the storage of a TensorImpl changed.
+  // Step 5) Handle ABA for TensorImpl.
```

Comment:
I really like this idea. I had a similar concern while writing a simplified logic for tensor identification/categorization.